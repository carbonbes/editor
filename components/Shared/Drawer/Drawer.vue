<template>
  <DrawerRoot
    v-bind="forwarded"
    v-model:open="open"
    v-model:active-snap-point="state.activeSnapPoint"
  >
    <slot />
  </DrawerRoot>
</template>

<script setup lang="ts">
import { DrawerRoot, type DrawerRootEmits, type DrawerRootProps } from 'vaul-vue'
import { useForwardPropsEmits } from 'reka-ui'
import { drawerContextKey } from '~/keys'

const props = defineProps<DrawerRootProps>()
const emits = defineEmits<DrawerRootEmits>()

const forwarded = useForwardPropsEmits(props, emits)

const open = defineModel<boolean>('open')

let state = reactive({
  snapPoints: props.snapPoints,
  activeSnapPoint: props.snapPoints?.[0] || 1,
})

type UpdateStateInputArgs = { open: boolean, snapPoints: number[], activeSnapPoint: number }

function updateState(args: UpdateStateInputArgs) {
  const { open } = args

  state = { ...args }
}

provide(drawerContextKey, { state: { ...state, open }, updateState })
</script>
