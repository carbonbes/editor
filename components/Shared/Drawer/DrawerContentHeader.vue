<template>
  <Flex items-center class="p-6">
    <p v-if="title" class="text-xl font-bold">
      {{ title }}
    </p>

    <Flex
      center
      as-child
      class="ml-auto p-1.5 rounded-full bg-gray-200/50"
    >
      <button @click="updateState({ open: false })">
        <Icon name="tabler:x" class="size-5" />
      </button>
    </Flex>
  </Flex>
</template>

<script setup lang="ts">
import Flex from '~/components/Shared/Flex.vue'
import { drawerContextKey } from '~/keys'

const { title } = defineProps<{ title?: string }>()

const { updateState } = inject(drawerContextKey)
</script>
