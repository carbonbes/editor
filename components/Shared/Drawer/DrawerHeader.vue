<template>
  <Flex items-center>
    <p v-if="title" class="text-xl font-bold">
      {{ title }}
    </p>

    <Flex
      center
      as-child
      class="ml-auto p-1.5 rounded-full bg-gray-200/50"
    >
      <DrawerClose>
        <Icon name="tabler:x" class="size-5" />
      </DrawerClose>
    </Flex>
  </Flex>
</template>

<script setup lang="ts">
import Flex from '~/components/Shared/Flex.vue'
import { DrawerClose } from 'vaul-vue'

const { title } = defineProps<{ title?: string }>()
</script>
