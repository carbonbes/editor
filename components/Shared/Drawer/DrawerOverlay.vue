<template>
  <DrawerOverlay class="fixed inset-0 bg-black/50" />
</template>

<script setup lang="ts">
import { DrawerOverlay } from 'vaul-vue'
</script>
