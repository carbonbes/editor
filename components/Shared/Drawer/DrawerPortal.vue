<template>
  <DrawerPortal to="#teleports">
    <slot />
  </DrawerPortal>
</template>

<script setup lang="ts">
import { DrawerPortal } from 'vaul-vue'
</script>
