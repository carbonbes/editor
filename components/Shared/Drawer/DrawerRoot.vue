<template>
  <DrawerRoot
    v-bind="forwarded"
    v-model:open="open"
    v-model:active-snap-point="activeSnapPoint"
    :snap-points
  >
    <slot />
  </DrawerRoot>
</template>

<script setup lang="ts">
import { createContext, useForwardPropsEmits } from 'reka-ui'
import {
  DrawerRoot,
  type DrawerRootEmits,
  type DrawerRootProps,
} from 'vaul-vue'

const props = defineProps<DrawerRootProps>()
const emits = defineEmits<DrawerRootEmits>()

const forwarded = useForwardPropsEmits(props, emits)

const open = defineModel<boolean>('open', { default: false })

const activeSnapPoint = ref(props.activeSnapPoint && 1)
const contentScrollOffset = ref(0)

provideDrawerContext({ open, activeSnapPoint, contentScrollOffset })
</script>

<script lang="ts">
export type DrawerContext = {
  open: Ref<boolean>
  activeSnapPoint: Ref<number>
  contentScrollOffset: Ref<number>
}

export const [injectDrawerContext, provideDrawerContext]
  = createContext<DrawerContext>('Drawer')
</script>
