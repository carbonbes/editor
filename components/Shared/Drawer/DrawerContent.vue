<template>
  <Primitive
    ref="drawerScrollableContent"
    as-child
    class="overflow-y-auto"
    :class="scrollableContentClasses"
  >
    <slot />
  </Primitive>
</template>

<script setup lang="ts">
import { Primitive } from 'reka-ui'
import { injectDrawerContext } from './Drawer.vue'
import { usePointerSwipe, useScroll } from '@vueuse/core'

const drawerScrollableContent = useTemplateRef<HTMLElement>('drawerScrollableContent')

const { activeSnapPoint, YScrollOffset } = injectDrawerContext()

const { isSwiping, direction: swipeDirection } = usePointerSwipe(
  drawerScrollableContent,
  { threshold: 0 },
)

const { y } = useScroll(drawerScrollableContent)

watch(y, (y) => {
  YScrollOffset.value = y
}, { immediate: true })

const scrollableContentClasses = computed(() => ({
  '!overflow-y-hidden':
    activeSnapPoint.value !== 1 ||
    (activeSnapPoint.value === 1 &&
      y.value === 0 &&
      isSwiping.value &&
      swipeDirection.value === 'down'),

  '!overflow-y-auto':
    activeSnapPoint.value === 1 &&
    y.value === 0 &&
    isSwiping.value &&
    swipeDirection.value === 'up',
}))
</script>
