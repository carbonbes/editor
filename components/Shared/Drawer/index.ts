import DrawerRoot from './DrawerRoot.vue'
import { DrawerTrigger } from 'vaul-vue'
import DrawerPortal from './DrawerPortal.vue'
import DrawerOverlay from './DrawerOverlay.vue'
import DrawerBody from './DrawerBody.vue'
import DrawerHeader from './DrawerHeader.vue'
import DrawerContent from './DrawerContent.vue'

export { DrawerRoot as Drawer, DrawerTrigger, DrawerPortal, DrawerOverlay, DrawerBody, DrawerHeader, DrawerContent }
