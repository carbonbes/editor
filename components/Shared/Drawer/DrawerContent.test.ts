import { describe, expect, test } from 'vitest'
import { DrawerContent } from '~/components/Shared/Drawer/index'
import mountDrawer from '~/tests/utils/mountDrawer'

describe('DrawerContent', () => {
  const template = `
    <DrawerBody>
      <DrawerContent>
        <slot />
      </DrawerContent>
    </DrawerBody>
  `

  test('should be defined', () => {
    expect(DrawerContent).toBeDefined()
  })

  test('should be displayed without the slot', async () => {
    const wrapper = await mountDrawer({ template })
    const content = wrapper.getComponent(DrawerContent)

    expect(content.html()).toMatchSnapshot()

    wrapper.unmount()
  })

  test('should be displayed with a slot', async () => {
    const Content = defineComponent({
      template: `
        <p>Test</p>
      `,
    })

    const wrapper = await mountDrawer({ template, slot: Content })
    const content = wrapper.getComponent(DrawerContent)

    expect(content.html()).toMatchSnapshot()

    wrapper.unmount()
  })
})
