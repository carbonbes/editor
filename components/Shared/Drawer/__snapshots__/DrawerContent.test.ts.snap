// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrawerContent > should be displayed with a slot 1`] = `
"<div class="overflow-y-auto">
  <p>Test</p>
</div>"
`;

exports[`DrawerContent > should be displayed without the slot 1`] = `"<div class="overflow-y-auto"></div>"`;
